src/ripple/app/misc/AMMHelpers.h

```diff
@@ -0,0 +1,224 @@
+ * @param tfee  trading fee in {0, 1000}
+ * 1 = 1/10bps or 0.001%, 1000 = 1%
+ */
+inline Number
+getFee(std::uint16_t tfee)
+{
+    return Number{tfee} / 100000;
+}
+
+/** Get fee multiplier (1 - tfee)
+ * @tfee trading fee in basis points
+ */
+inline Number
+feeMult(std::uint16_t tfee)
+{
+    return 1 - getFee(tfee);
+}
+
+/** Get fee multiplier (1 - tfee / 2)
+ * @tfee trading fee in basis points
+ */
+inline Number
+feeMultHalf(std::uint16_t tfee)
+{
+    return 1 - getFee(tfee) / 2;
+}
+
+/** Calculate LP Tokens given asset's deposit amount.
+ * @param asset1Balance current AMM asset1 balance
+ * @param asset1Deposit requested asset1 deposit amount
+ * @param lpTokensBalance LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return tokens
+ */
+STAmount
+lpTokensIn(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Deposit,
+    STAmount const& lpTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate asset deposit given LP Tokens.
+ * @param asset1Balance current AMM asset1 balance
+ * @param lpTokensBalance LP Tokens balance
+ * @param ammTokensBalance AMM LPT balance
+ * @param tfee trading fee in basis points
+ * @return
+ */
+STAmount
+ammAssetIn(
+    STAmount const& asset1Balance,
+    STAmount const& lpTokensBalance,
+    STAmount const& ammTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate LP Tokens given asset's withdraw amount. Return 0
+ * if can't calculate.
+ * @param asset1Balance current AMM asset1 balance
+ * @param asset1Withdraw requested asset1 withdraw amount
+ * @param lpTokensBalance LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return tokens out amount
+ */
+STAmount
+lpTokensOut(
+    STAmount const& asset1Balance,
+    STAmount const& asset1Withdraw,
+    STAmount const& lpTokensBalance,
+    std::uint16_t tfee);
+
+/** Calculate asset withdrawal by tokens
+ * @param assetBalance balance of the asset being withdrawn
+ * @param lptAMMBalance total AMM Tokens balance
+ * @param lpTokens LP Tokens balance
+ * @param tfee trading fee in basis points
+ * @return calculated asset amount
+ */
+STAmount
+withdrawByTokens(
+    STAmount const& assetBalance,
+    STAmount const& lptAMMBalance,
+    STAmount const& lpTokens,
+    std::uint32_t tfee);
+
+/** Find in/out amounts to change the spot price quality to the requested
+ * quality. Implements AMM Swap equation (11) and calls swapAssetIn
+ * to find the required offer size to change the spot price quality.
+ * @param pool AMM pool balances
+ * @param quality requested quality
+ * @param tfee trading fee in basis points
+ * @return seated in/out amounts if the quality can be changed
+ */
+template <typename TIn, typename TOut>
+std::optional<TAmounts<TIn, TOut>>
+changeSpotPriceQuality(
+    TAmounts<TIn, TOut> const& pool,
+    Quality const& quality,
+    std::uint32_t tfee)
+{
+    if (auto const nTakerPays =
+            (root2(pool.in * pool.out * quality.rate()) - pool.in);
+        nTakerPays > 0)
```

Comment:
I want to walk through this one, because there's a lot happening in this equation (11), before it's even converted to code.
To start with, given equal weights, equation 11 becomes
<p>$$\Delta_{B} = \Gamma_{B} * \left[ \left( \frac{SP'^{B}_{A}}{SP^{B}_{A}} \right) ^ \frac{1}{2} -1 \right] $$</p>
Equation II, given equal weights, becomes
<p>$$SP^{B}_{A} = \frac{\Gamma_{B}}{\Gamma_{A}} * \frac{1}{\left( 1 - TFee \right)}$$</p>
Now, assuming I'm understanding the variables correctly, we can make the following substitutions:
<p>$$nTakerPays = \Delta_{B}$$</p>
<p>$$pool.in = \Gamma_{B}$$</p>
<p>$$pool.out = \Gamma_{A}$$</p>
<p>$$quality.rate() = SP'^{B}_{A}$$</p>
<p>$$feeMult(tfee) = \left( 1 - TFee \right)$$</p>
That results in:
<p>$$nTakerPays = pool.in * \left[ \left( \frac{quality.rate()}{\frac{pool.in}{pool.out} * \frac{1}{feeMult(tFee)}} \right) ^ \frac{1}{2} -1 \right] $$</p>
Which simplifies to:
<p>$$nTakerPays = pool.in * \left[ \left( \frac{pool.out * quality.rate() * feeMult(tFee)}{pool.in} \right) ^ \frac{1}{2} -1 \right] $$</p>
Distributing `pool.in` , we get:
<p>$$nTakerPays = \left( pool.in^2 * \frac{pool.out * quality.rate() * feeMult(tFee)}{pool.in} \right) ^ \frac{1}{2} - pool.in$$</p>
And finally:
<p>$$nTakerPays = \left( pool.in * pool.out * quality.rate() * feeMult(tFee) \right) ^ \frac{1}{2} - pool.in$$</p>

Translated to code, that becomes
```
auto const nTakerPays = root2( pool.in * pool.out * quality.rate() * feeMult(tFee) ) - pool.in;
```
Does that sound right? If so, then you need to add the `* feeMult(tFee)` up there.

Reply:
You are right!

Reply:
Woo hoo!!! Yes! Math!

Reply:
Actually, after giving it some thought/testing this needs to be updated as following:

Spot price quality (SPQ) has to be changed when there s a CLOB offer for the same token pair as AMM and its quality **q** is worse than AMM's SPQ. In this case we need to generate a synthetic AMM offer so that the updated pool SPQ is equal to **q** and the AMM offer quality is better than **q**. To change the AMM pool to a new SPQ **q**, we need to find **i** (taker pays) and **o** (taker gets) such that given the pool composition **I** (pool gets) and **O** (pool pays), and the trading fee **tfee**:

$$q = {O - o \over I + i} \tag{1}$$

Where **o** is calculated with the swap-in formula:

$$ o = O * (1 - {I \over I + i * (1 - tfee)}) \tag{2}$$

Let $f = 1 - tfee$. Then simplifying (2) we get:

$$ o = {O * i * f \over I + i*f} \tag{3} $$

Substituting $o$ in (1) with (3) and simplifying:

$$ q = {O*I \over (I+i)*(I+i*f)} \tag{4}$$

Or:

$$ f*i^2 + (I*f + I)*i + I^2 - {I*O \over q} = 0 \tag{5}$$

This can be solved with the quadratic equation solution:

$$ {-b + \sqrt {b^2 - 4 * a * c} \over 2*a} \tag{6}$$

Where $a=(1-tfee)$, $b=I*(1-tfee) + I$, $c={I*O \over q}$

Solving (5) results in the new SPQ matching the requested quality **q**. However, if the trading fee is too high then the synthetic AMM offer quality could be less than **q**, which results in a worse trade from the taker prospective. Therefore, there needs to be a constraint on **i**:

$$ {o \over i} >= q \tag{7}$$

Substituting $o$ with (3) and simplifying:

$$ i <= {{O \over q} - {I \over f}} \tag{8}$$

To generate the synthetic AMM offer we first solve (5). If **i** meets the constraint (8) then (3) is used to find **o**. Otherwise **i** is set to ${{O \over q} - {I \over f}}$ and then **o** is found with (3).